blockchain/transaction_confirmers.go

```diff
@@ -298,36 +299,84 @@ func (e *EthereumClient) GetHeaderSubscriptions() map[string]HeaderEventSubscrip
 	return newMap
 }
 
-// subscribeToNewHeaders
+// subscribeToNewHeaders kicks of the primary header subscription for the test loop
 func (e *EthereumClient) subscribeToNewHeaders() error {
 	headerChannel := make(chan *SafeEVMHeader)
 	subscription, err := e.SubscribeNewHeaders(context.Background(), headerChannel)
 	if err != nil {
 		return err
 	}
-	defer subscription.Unsubscribe()
-
 	log.Info().Str("Network", e.NetworkConfig.Name).Msg("Subscribed to new block headers")
 
+	go e.headerSubscriptionLoop(subscription, headerChannel)
+	return nil
+}
+
+// headerSubscriptionLoop receives new headers, and handles subscription errors when they pop up
+func (e *EthereumClient) headerSubscriptionLoop(subscription ethereum.Subscription, headerChannel chan *SafeEVMHeader) {
+	lastHeaderNumber := uint64(0)
 	for {
 		select {
-		case err := <-subscription.Err():
-			log.Error().Err(err).Msg("Error while subscribed to new headers, restarting subscription")
+		case err := <-subscription.Err(): // Most subscription errors are temporary RPC downtime, so let's poll to resubscribe
+			log.Error().Err(err).Msg("Error while subscribed to new headers, likely RPC errors")
 			subscription.Unsubscribe()
 
-			subscription, err = e.SubscribeNewHeaders(context.Background(), headerChannel)
-			if err != nil {
-				log.Error().Err(err).Msg("Failed to resubscribe to new headers")
-				return err
+			rpcDegradedTime, rpcDegradedNotifyTime := time.Now(), time.Now()
+			timeout, ticker := time.NewTimer(e.NetworkConfig.Timeout.Duration), time.NewTicker(time.Second)
+		reSubLoop:
+			for { // Loop to resubscribe to new headers,
+				select {
+				case <-timeout.C: // Timeout waiting for RPC to come back up
+					log.Error().Dur("Time waiting", time.Since(rpcDegradedTime)).Msg("RPC connection still down, timed out waiting for it to come back up")
+					e.Client.Close()
+					return
+				case <-ticker.C: // Poll the RPC connection to see if it's back up
+					if time.Since(rpcDegradedNotifyTime) >= time.Second*15 { // Periodically inform that we're still waiting for RPC to come back up
+						log.Warn().Dur("Time waiting", time.Since(rpcDegradedTime)).Msg("RPC connection still down, waiting for it to come back up")
+						rpcDegradedNotifyTime = time.Now()
+					}
+					subscription, err = e.SubscribeNewHeaders(context.Background(), headerChannel)
+					if err == nil { // No error on resubscription, RPC connection restored, back to regularly scheduled programming
```

Comment:
do we need to handle the else part?

Reply:
We do, we just log a trace that we failed that attempt and go back to the beginning of the loop and wait for the next tick.